staker/block_validator.go

```diff
@@ -490,527 +426,436 @@ func (v *BlockValidator) SetCurrentWasmModuleRoot(hash common.Hash) error {
-				if writeErr != nil {
-					log.Warn("failed to write validation debugging info", "err", err)
+			var wasmRoots []common.Hash
+			for i, run := range validationStatus.Runs {
+				if !run.Ready() {
+					log.Trace("advanceValidations: validation not ready", "pos", pos, "run", i)
+					continue validatiosLoop
+				}
+				wasmRoots = append(wasmRoots, run.WasmModuleRoot())
+				runEnd, err := run.Current()
+				if err == nil && runEnd != validationStatus.Entry.End {
+					err = fmt.Errorf("validation failed: expected %v got %v", validationStatus.Entry.End, runEnd)
+					writeErr := v.writeToFile(validationStatus.Entry, run.WasmModuleRoot())
+					if writeErr != nil {
+						log.Warn("failed to write debug results file", "err", writeErr)
+					}
 				}
-				v.possiblyFatal(err)
+				if err != nil {
+					validatorFailedValidationsCounter.Inc(1)
+					v.possiblyFatal(err)
+					return &pos, nil // if not fatal - retry
+				}
+				validatorValidValidationsCounter.Inc(1)
 			}
+			v.lastValidGS = validationStatus.Entry.End
+			go v.recorder.MarkValid(pos, v.lastValidGS.BlockHash)
+			err := v.writeLastValidatedToDb(validationStatus.Entry.End, wasmRoots)
 			if err != nil {
-				v.possiblyFatal(err)
-				validationStatus.setStatus(Failed)
-				validatorFailedValidationsCounter.Inc(1)
-				validatorPendingValidationsGauge.Dec(1)
-				return
+				log.Error("failed writing new validated to database", "pos", pos, "err", err)
+			}
+			atomicStorePos(&v.validatedA, pos+1)
+			nonBlockingTriger(v.createNodesChan)
+			nonBlockingTriger(v.sendRecordChan)
+			validatorMsgCountValidatedGauge.Update(int64(pos + 1))
+			if v.testingProgressMadeChan != nil {
+				nonBlockingTriger(v.testingProgressMadeChan)
 			}
+			v.maybePrintNewlyValid()
+			continue
 		}
-		for _, run := range validationStatus.Runs {
-			run.Cancel()
+		if room == 0 {
+			log.Trace("advanceValidations: no more room", "pos", pos)
+			return nil, nil
 		}
-		validationStatus.replaceStatus(ValidationSent, Valid)
-		validatorValidValidationsCounter.Inc(1)
-		validatorPendingValidationsGauge.Dec(1)
-		v.triggerSendValidations()
-		earliestBatchKept := atomic.LoadUint64(&v.earliestBatchKept)
-		seqMsgNr := validationEntry.StartPosition.BatchNumber
-		if earliestBatchKept < seqMsgNr {
-			for batch := earliestBatchKept; batch < seqMsgNr; batch++ {
-				v.sequencerBatches.Delete(batch)
+		if currentStatus == Prepared {
+			replaced := validationStatus.replaceStatus(Prepared, SendingValidation)
+			if !replaced {
+				v.possiblyFatal(errors.New("failed to set SendingValidation status"))
 			}
-			atomic.StoreUint64(&v.earliestBatchKept, seqMsgNr)
+			v.LaunchThread(func(ctx context.Context) {
+				validationCtx, cancel := context.WithCancel(ctx)
+				defer cancel()
+				validationStatus.Cancel = cancel
+				input, err := validationStatus.Entry.ToInput()
+				if err != nil && validationCtx.Err() == nil {
+					v.possiblyFatal(fmt.Errorf("%w: error preparing validation", err))
+					return
```

Comment:
Shouldn't we instead set the status to `RecordFailed` here? I think this error gets hit when we fail to fetch a DAS batch, and we don't want to exit due to that. As-is I don't think `RecordFailed` is used.

Reply:
Alternatively, maybe we could set the status back to Prepared? Either way we should log a warning here.

Reply:
this is not a recording error. recording errors are handled in sendRecord.
This in an error in Entry.ToInput() - transalting a ready entro to json request.
These are not expected. Currently the only codepath I see is if the validationStatus is marked Prepared while the underline Entry is not marked Ready.
If we ever see that in production we might reconsider but I think shutting down is the correct approach here.

Reply:
:+1: I must have confused this with a similar function pre-execution-split